### ISSUE TITLE
Stream ending with BadState when navigating back and forth quickly

### ISSUE THREAD

**Issue by AxelBlaz3** [rrousselGit/riverpod#3485]
**Describe the bug**
I have an app where there is a button in the home page. Upon clicking the button, it navigates me to accounts list page which are retrieved from a local database query stream. When I go to accounts page and come back, everything works fine. However, if I **immediately** go to accounts page as soon as I am back to home page, it throws me `BadState: Stream has already been listened to` error. Upon investigating, I found that it takes about a second to cleanup the provider state. Meaning, if I navigate **do not** navigate quickly to accounts page from home, it works fine. 

I tried it with two different databases - `Isar` and `ObjectBox`. Both of them were having the same behaviour. If I remove the provider and manually get the stream, it works fine because the widget lifecycle disposes of the stream as soon as I go back (pop the navigation stack).

**To Reproduce**
I am attaching a sample project with instructions on the home page to reproduce the bug.
https://github.com/AxelBlaz3/Riverpod-StreamBuilder-Bug

**Expected behavior**
Riverpod should clean up the provider state immediately after I pop the current page because there are no listeners after popping the current page. However, it takes about a second to clean things up which is not what I am expecting (but correct me if I am mistaken).

**Comment by AxelBlaz3** [comment]
I am attaching the demo video. If you look carefully you will see that I get a bad stream error only when I quickly navigate to accounts page from home page (I got to home page form accounts page by back gesture on my device). It works fine on normal navigation speed.

https://github.com/rrousselGit/riverpod/assets/28803728/979c65ad-3656-418e-bef6-d9a7c72c1dbe

Just as a note, I have `AccountsUiState` class with only `Stream<List<Account>>` because I wanted it to be minimal. In real project, I have more than just a stream.

**Comment by RandalSchwartz** [comment]
Just filling in, it looks like the stream is being listened twice, which a singleuser stream cannot.  But the puzzle is why doesn't the provider fully recycle during the quick back and forth, which would have created a new stream?

**Comment by AxelBlaz3** [comment]
@RandalSchwartz exactly, I am curious why the provider does not fully recycle during quick back and forth.

**Comment by rrousselGit** [comment]
There's too much code here. I can't realistically investigate a whole project. Please create a minimal reproducible example.

But my guess is, this is most likely a mistake on your end. I'd suggest asking for help on Discord

**Comment by AxelBlaz3** [comment]
@rrousselGit thanks for letting me know about it. I apologize for having a bit more code for a minimal sample. I hope the new code changes are simple to catch and understand. Have a look at it at your convenience.

Also, I have tried asking for help on Discord before I filed an issue. I could be a mistake from my end but I can't figure out where I am going wrong with this.

**Comment by AxelBlaz3** [comment]
As a follow-up, I noticed that when I converted AccountsPage to `ConsumerStatefulWidget` and logged `initState` and `dispose`, `dispose` is being called with a delay when I navigate back.

**Comment by AxelBlaz3** [comment]
I recently pushed a new branch called `working`, which works flawlessly and does not use riverpod to inject dependencies. It works fine regardless of how quick I navigate. The `dispose()` delay seems to be normal and it is how flutter handles internally from my observation. 

As a note, the file `providers.dart` in `working` branch has `Future`s for global state.

**Comment by AxelBlaz3** [comment]
It seems like I have to `await for` my stream in notifier's `build()` and then `yield` the `List<Account>`. This solves the problem but curious why using a `StreamBuilder` is not working.

**Comment by lucavenir** [comment]
I am *unable* to reproduce this issue.

I tried:

- Using the [linked repository](https://github.com/AxelBlaz3/Riverpod-StreamBuilder-Bug) - no exception was thrown, nor logs were shown
- Tried writing a test without Flutter Widgets
- Tried writing a test with Flutter Widgets

None of these led to the problem.

@AxelBlaz3 are you absolutely positive about this?

<details><summary>Test without widgets</summary>

```dart
import 'package:riverpod/riverpod.dart';
import 'package:test/test.dart';

void main() {
  test('stream handling', () async {
    final container = ProviderContainer();
    addTearDown(container.dispose);

    var sub = container.listen(accountsListProvider.future, (_, __) {});
    final stream = await sub.read();
    expectLater(stream.first, completion(0));
    await Future.delayed(Duration(milliseconds: 100));
    sub.close();
    sub = container.listen(accountsListProvider.future, (_, __) {});
    final j = sub.read();
    expectLater(j, completes);
  });
}

final accountsStreamProvider = FutureProvider.autoDispose<Stream<int>>((ref) {
  return Future.delayed(
    Duration(milliseconds: 400),
    () => Stream.fromIterable([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]),
  );
});

final accountsListProvider =
    AutoDisposeAsyncNotifierProvider<AccountsListNotifier, Stream<int>>(
        AccountsListNotifier.new);

class AccountsListNotifier extends AutoDisposeAsyncNotifier<Stream<int>> {
  @override
  Future<Stream<int>> build() async {
    return ref.watch(accountsStreamProvider.future);
  }
}
```
</details> 

<details><summary>Test with widgets</summary>

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('???', (tester) async {
    await tester.pumpWidget(
      ProviderScope(
        child: SomeWidget(),
      ),
    );
    await tester.tap(find.text("push me"));
    await tester.pumpAndSettle();
    await tester.tap(find.byType(IconButton));
    await tester.pump();
    await tester.tap(find.text("push me"));
  });
}

class SomeWidget extends StatelessWidget {
  const SomeWidget({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: Builder(builder: (context) {
            return TextButton(
              onPressed: () {
                Navigator.of(context).push(MaterialPageRoute(
                  builder: (context) => StreamWidget(),
                ));
              },
              child: Text("push me"),
            );
          }),
        ),
      ),
    );
  }
}

class StreamWidget extends ConsumerWidget {
  const StreamWidget({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final stream = ref.watch(accountsListProvider);

    return Scaffold(
      appBar: AppBar(
        title: Text("StreamWidget"),
      ),
      body: switch (stream) {
        AsyncData(:final value) => StreamBuilder(
            stream: value,
            builder: (context, snapshot) {
              return Text(snapshot.data.toString());
            },
          ),
        AsyncError() => Text("Error"),
        _ => Text("Loading")
      },
    );
  }
}

final accountsStreamProvider = FutureProvider.autoDispose<Stream<int>>((ref) {
  return Future.delayed(
    Durations.medium4,
    () => Stream.fromIterable([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]),
  );
});

final accountsListProvider =
    AutoDisposeAsyncNotifierProvider<AccountsListNotifier, Stream<int>>(
        AccountsListNotifier.new);

class AccountsListNotifier extends AutoDisposeAsyncNotifier<Stream<int>> {
  @override
  Future<Stream<int>> build() async {
    return ref.watch(accountsStreamProvider.future);
  }
}
```
</details>

**Comment by rrousselGit** [comment]
@lucavenir Don't remove the `bug` tag in those cases. Just add `question`.
